lib/cog/command/pipeline/destination.ex

```diff
@@ -0,0 +1,123 @@
+defmodule Cog.Command.Pipeline.Destination do
+
+  @type output_level :: :full | :status_only
+
+  @type t :: %__MODULE__{raw: String.t,
+                         output_level: output_level,
+                         adapter: String.t,
+                         room: term}
+  defstruct [raw: nil,
+             output_level: :full,
+             adapter: nil,
+             room: nil]
+  use Adz
+
+  @doc """
+  Given a list of raw pipeline destinations, resolve them all to the
+  appropriate adapter-specific destinations, or return error tuples
+  for each that were invalid.
+  """
+  def process(raw_destinations, sender, origin_room, origin_adapter) do
+    raw_destinations
+    |> Enum.map(&make_destination/1)
+    |> maybe_add_origin(origin_adapter)
+    |> resolve(sender, origin_room, origin_adapter)
+  end
+
+  ########################################################################
+
+  defp make_destination(raw),
+    do: %__MODULE__{raw: raw}
+
+  # If no destinations are explicitly provided, we return the full
+  # output to the same place the request originated from. For chat
+  # adapters, this is the same chat room the pipeline was invoked
+  # from. For non-chat adapters (e.g., HTTP request), it is the source
+  # of the pipeline request (e.g., the HTTP request --- the output
+  # will be the body of the HTTP response).
+  #
+  # If destinations _are_ provided, but the origin is not one of them
+  # (denoted by the reserved target "here"), we add a status-only
+  # "here" destination, but only if the request came from a non-chat
+  # adapter. We do this because, e.g., HTTP requests need a response
+  # of some kind, and we want to indicate whether or not the pipeline
+  # succeeded.
+  #
+  # If the origin _is_ one of the destinations, we do nothing special;
+  # all explicitly listed destinations receive full output.
+  defp maybe_add_origin([], _origin_adapter),
+    do: [make_destination("here")]
+  defp maybe_add_origin(destinations, origin_adapter) do
+    if !originator_is_destination?(destinations) && !origin_adapter.chat_adapter? do
```

Comment:
Might be nice to invert the boolean expression here to `originator_is_destination?(destinations) || origin_adapter.chat_adapter?`